### PR TITLE
Release/v0.4.0

### LINKED ISSUE #21: README.rst needs a code coverage badge
It would be nice to track code coverage in ``README.rst``.

The project:  
https://github.com/dbrgn/coverage-badge

Seems promising, but we will need a place to host the badge after it has been generated.

Hosting in a gist seems like a possibility.

### LINKED ISSUE #31: convert unnamed outputs to named outputs
it would make things easier for the devs if there was a unified output type; perhaps a function converting unnamed outputs to named outputs would solve this

### LINKED ISSUE #82: Tracking iteration count / function evaluation count in a simulation
I have a simulation

```python
my_moop.addSimulation({ 'name': "my_opt", 'm': 2, 'sim_func': evaluate_iteration, ... )
```

Where `evaluate_iteration()` needs information about its iteration id / evaluation id, to do things like create a subdirectory within which to work (e.g., to avoid temporary file name-clashing):

```python
test_name = f"evaluation_{eval_id}"
subdir = os.path.join( top_wd, test_name )
if not os.path.exists( subdir ):
    os.makedirs( subdir )
os.chdir( subdir )
```

I was wondering how I might do that with ParMOO... did I miss something in the documentation?

### LINKED ISSUE #93: Update release proc to include coverage
As noted in #92, we may want to have separate badges in `main` and `develop` for coverage.

Probably there is a good way to do this using actions, but `main` should at least track the `main` coverage.

### PR DESCRIPTION
**Pull request by thchang**
Major interface-breaking refactor of core `MOOP` builder and libraries plus
minor bug-fixes.

Major changes:

 - Under the hood, ParMOO now uses `jax[cpu]` for all gradient evaluations,
   constraint evaluations, and function evaluations (but not simulation
   evaluations) on the critical path -- this means that objective and
   constraint functions can be just-in-time (jit) compiled, which can give
   massive performance improvements.  However, not all Python features are
   supported by `jax.jit()` so users must be careful to write their objective
   and constraint functions mindfully
 - In order for jax to be effective, we have updated ParMOO's interfaces to
   avoid using optional arguments -- this affects the objective and constraint
   function interfaces, and a separate gradient function must be provided by
   the user (as opposed to an optional `der` argument, as previously used)
 - In order to make it easier to support mixed variable types in jax and for
   future maintainability, all design variables have been replaced by a library
   of design variable `Embedder` classes (corresponding ABCs also added to the
   `structs.py`)
 - In order to make ParMOO more maintainable and for jax to work smoothly, we
   have dropped support for unnamed variables (Closes #31)
 - All `SurrogateFunction` and `AcqusitionFunction` libraries have been updated
   to be more jax-friendly
 - The `SurrogateOptimizer` class has been refactored to include a callback to
   observe simulation evaluation results.  It has also been given almost full
   control over when model improvement steps are called to make implementing
   most DFO methods easier
 - The pattern search family of optimizers has been greatly improved.
 - `PyDOE` has been dropped since most relevant DOEs now appear in the newly
   added `scipy.stats.qmc` module
 - Switching to `PyDOE` required us to change how the numpy random seed is set.
   A random seed object is now passed as a hyperparameter to ParMOO and
   propagated to all libraries, which is the recommended way anyway
 - Updated docs to reflect above changes

Style changes:

 - Overhauled some of the unit tests
 - Style fixes throughout
 - Renamed several methods and classes to have a consistent naming convention

Interface breaking:

 - The `Embedder` class is now used to define custom design variables (see
   Major changes)
 - jax is now used to evaluate gradients in the `SurrogateOptimizer` class (see
   Major changes) -- this alone shouldn't break interface most use-cases, but
   may lead to decreased performance if not careful (see notes on achieving
   good performance in jax in the docs).  Additionally, jax defaults to single
   precision so double-precision must be set manually using a `jax.config`
   command (see docs)
 - `SurrogateFunction`, `SurrogateOptimizer`, and `AcquisitionFunction`
   interfaces have changed (only affects users using custom methods)
 - The random seed must now be set using a numpy random seed object (see Major
   changes and examples in the docs)
 - We no longer support unnamed variables (see Major changes)
 - Gradient functions are now provided via an additional key in the dictionary,
   and cannot be set using an optional argument (see Major changes and examples
   in the updated docs)
 - Many library functions, methods, and classes have been renamed for
   consistency.  In some cases, the old names remain as aliases for backward
   compatibility.  See Style changes

New features:

 - ParMOO now supports jax for autograd (see Major changes above)
 - `SurrogateOptimizer` is now notified of the results of each simulation
   evaluation. (This allows checks for sufficient improvements)
 - Numerous new `AcquisitionFunction` types added.
 - Added an option to create a private workdir for each libEnsemble thread
   (Closes #82)

Minor changes:

 - Updates to support `numpy 2.0`
 - Added a code coverage badge and updated the release process to reflect the
   extra steps needed to make this work (Closes #21 , Closes #93)

Requirements:

 - Added `jax[cpu]` to list of requirements
 - Removed `pyDOE` from list of requirements, in favor of `scipy.stats` (added
   in `scipy 1.10.0`)
 - Released the lock on `libensemble` version (from Release 0.3.1)
 - Updated all version requirements to be new enough to support `numpy 2.0`

### CODE REVIEW

**Comment by wildsm** [comment]
@thchang  I think we should first pull in #97 (can be into this branch if not develop)

**Comment by wildsm** [comment]
For reference, the diffs between this release branch and develop are:
https://github.com/parmoo/parmoo/compare/develop...release/v0.4.0

**Comment by thchang** [comment]
@wildsm  --  I added spellcheck to the release process.

However, when I run it, I get the following error:

```
ERROR: CHANGELOG.rst -- ('EOF in multi-line statement', (311, 0))
!!!Spelling check failed!!!
```

Have you seen this before?  I'm assuming it's thrown off by some unclosed parentheses or something similar, but I can't figure out where in the file

**Comment by wildsm** [comment]
It did not like the 

`(Closes #X)`

statements. Making them consistent with the rest of the CHANGELOG fixed the problem, please try now.

**Comment by wildsm** [comment]
Note that `du -h` shows  `8.6M	./docs/img` (most of which is the movie).

If ever we needed a lighterweight parmoo, could send that image to the solver farm

**Comment by thchang** [comment]
Great, can we also maybe do something about the spellcheck action that is failing?  I think this happens because we have the `.spellcheck.yaml` file in the workflows directory, and the job "fails automatically" because it is invalid.

It doesn't show up in the PR, but I see a lot of failed actions in the actions tab and get failure notification emails.

@wildsm

**Comment by wildsm** [comment]
> Great, can we also maybe do something about the spellcheck action that is failing? I think this happens because we have the `.spellcheck.yaml` file in the workflows directory, and the job "fails automatically" because it is invalid.
> 
> It doesn't show up in the PR, but I see a lot of failed actions in the actions tab and get failure notification emails.
> 
> @wildsm

I have now addressed this by creating a github action that successfully runs. 
I am still including the manual pyspelling line in the release-proc since it is helpful breadcumbs for someone trying to update or add to CI/github actions/the spelling workflow overall. Please update release-proc if you feel differently @thchang .

**Comment by wildsm** [comment]
Latest checks have almost passed and I have completed my review.
Note that the badge is not rendering correctly on this branch.

From https://coveralls.io/github/parmoo/parmoo?branch=main
I believe that this could work when this branch is merged, but this needs to be checked before a release is tagged @thchang

**Comment by wildsm** [comment]
Tests have passed.

I've created #101 to track release progress and added a temporary note to check the coveralls.

**Review by wildsm — APPROVED** [review]
Review complete, I suggest @thchang take a look at things that changed this afternoon

**Comment by thchang** [comment]
Looks good, I'm going to merge and I'll check to see if the coverage branch target resolves!

🥳
